src/diverse_seq/cluster.py

```diff
@@ -0,0 +1,618 @@
+
+        # Compute sketches in parallel
+        futures_to_idx = {
+            self._executor.submit(
+                mash_sketch,
+                record,
+                self._k,
+                self._sketch_size,
+                self._num_states,
+                canonical=self._canonical,
+            ): i
+            for i, record in enumerate(records)
+        }
+
+        for future in as_completed(list(futures_to_idx)):
+            idx = futures_to_idx[future]
+            bottom_sketches[idx] = future.result()
+
+            if self._with_progress:
+                self._progress.update(sketch_task, advance=1)
+
+        return bottom_sketches
+
+
+def get_ordered_records(
+    seq_store: HDF5DataStore,
+    seq_names: Sequence[str],
+) -> list[DataMember]:
+    """Returns ordered data store records given a seqeunce
+    of sequence names.
+
+    Parameters
+    ----------
+    seq_store : HDF5DataStore
+        The data store to load from.
+    seq_names : Sequence[str]
+        The ordered sequence names.
+
+    Returns
+    -------
+    list[DataMember]
+        Ordered data member records.
+    """
+    records = {m.unique_id: m for m in seq_store.completed if m.unique_id in seq_names}
+    return [records[name] for name in seq_names]
+
+
+def compute_chunk_distances(
+    start_idx: int,
+    stride: int,
+    sketches: list[BottomSketch],
+    k: int,
+    sketch_size: int,
+) -> tuple[int, dok_matrix]:
+    """Find a subset of pairwise distances between sketches.
+
+    Finds pairwise distances for the row with the given start index
+    and each stride after that.
+
+    Only finds lower triangular portion.
+
+    Parameters
+    ----------
+    start_idx : int
+        Start index for distance calculation.
+    stride : int
+        Index increment.
+    sketches : list[BottomSketch]
+        Sketches for pairwise distances.
+    k : int
+        kmer size.
+    sketch_size : int
+        Size of the sketches.
+
+    Returns
+    -------
+    tuple[int, dok_matrix]
+        The start index, and the calculated pairwise distances.
+    """
+    distances_chunk = dok_matrix((len(sketches), len(sketches)))
+    for i in range(start_idx, len(sketches), stride):
+        for j in range(i):
+            distance = compute_mash_distance(sketches[i], sketches[j], k, sketch_size)
+            distances_chunk[i, j] = distance
+    return start_idx, distances_chunk
+
+
+def mash_sketch(
+    record: DataMember,
+    k: int,
+    sketch_size: int,
+    num_states: int,
+    *,
+    canonical: bool,
+) -> BottomSketch:
+    """Find the mash sketch for a sequence record.
+
+    Parameters
+    ----------
+    record : DataMember
+        The sequence record to find the sketch for.
+    k : int
+        kmer size.
+    sketch_size : int
+        Size of the sketch.
+    num_states : int
+        Number of possible states (e.g. GCAT gives 4 for DNA).
+    canonical : bool
+        Whether to use the canonical representation of kmers.
```

Comment:
elsewhere in diverse_seqs, the word `canonical` means the non-ambiguous members of a molecular type, i.e. no ambiguity characters. What does `canonical` representation of k-mers mean?